tests/unit/tls_test.cc

```diff
@@ -201,6 +202,125 @@ SEASTAR_TEST_CASE(test_x509_client_with_priority_strings_fail) {
 }
 #endif // SEASTAR_TESTING_WITH_NETWORKING
 
+template <typename T, typename Func>
+auto with_https_server_and(T&& v, Func&& func) {
```

Comment:
Can we detemplate it?

 - make the caller capture `v` itself (it can use SEASTAR_THREAD_TEST_CASE if it wants a stable address)
 - change Func to std::function<future<> (socket_address sa, std::string cert)>
 - change the return type to future<>

If not, then the parameters must be constrained as a way to tell the reader what to expect.

Reply:
sure. now that `SEASTAR_THREAD_TEST_CASE` is used for adding the new tests, i take the opportunity to refactor `with_https_server_and()` to a class for better readability. it looks more closer to a coroutinized version.